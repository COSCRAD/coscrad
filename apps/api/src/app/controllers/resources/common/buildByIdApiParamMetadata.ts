export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
