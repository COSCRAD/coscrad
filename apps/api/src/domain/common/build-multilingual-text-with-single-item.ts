import { LanguageCode, MultilingualTextItemRole } from '@coscrad/api-interfaces';
import { MultilingualText, MultilingualTextItem } from './entities/multilingual-text';

/**
 * Sometimes there is a property that has a single piece of text. Typically,
 * this is in a context where only English text appears in the data (e.g., in
 * the research subdomain). This utility helps you convert such text to a proper
 * `MultilingualText` instance.
 *
 * @param text
 * @param languageCode
 * @returns a MultilingualText instance with a single item
 */
export const buildMultilingualTextWithSingleItem = (
    text: string,
    languageCode: LanguageCode = LanguageCode.English
) =>
    new MultilingualText({
        items: [
            new MultilingualTextItem({
                languageCode,
                // If there is only one text item, it must be the original
                role: MultilingualTextItemRole.original,
                text,
            }),
        ],
    });
