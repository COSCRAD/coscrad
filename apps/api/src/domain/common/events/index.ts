export * from './build-event-handler-maps';
export * from './coscrad-event-factory';
export * from './coscrad-event-sourced-aggregate-factory';
export * from './coscrad-event-union';
export * from './coscrad-event.decorator';
export * from './event.module';
