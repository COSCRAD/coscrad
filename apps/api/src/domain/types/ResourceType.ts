import { UuidDocument } from '../../lib/id-generation/types/UuidDocument';
import { IBibliographicReference } from '../models/bibliographic-reference/interfaces/bibliographic-reference.interface';
import { Book } from '../models/book/entities/book.entity';
import { Category } from '../models/categories/entities/category.entity';
import { EdgeConnection } from '../models/context/edge-connection.entity';
import { MediaItem } from '../models/media-item/entities/media-item.entity';
import { Photograph } from '../models/photograph/entities/photograph.entity';
import { Song } from '../models/song/song.entity';
import { ISpatialFeature } from '../models/spatial-feature/ISpatialFeature';
import { Tag } from '../models/tag/tag.entity';
import { Term } from '../models/term/entities/term.entity';
import { TranscribedAudio } from '../models/transcribed-audio/entities/transcribed-audio.entity';
import { CoscradUserGroup } from '../models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUser } from '../models/user-management/user/entities/user/coscrad-user.entity';
import { VocabularyList } from '../models/vocabulary-list/entities/vocabulary-list.entity';
import { AggregateId } from './AggregateId';

export enum ResourceType {
    term = 'term',
    vocabularyList = 'vocabularyList',
    transcribedAudio = 'transcribedAudio',
    book = 'book',
    photograph = 'photograph',
    spatialFeature = 'spatialFeature',
    bibliographicReference = 'bibliographicReference',
    song = 'song',
    mediaItem = 'mediaItem',
}

export const isResourceType = (input: unknown): input is ResourceType =>
    Object.values(ResourceType).includes(input as ResourceType);

// We should use this for type inference a few places.
export type ResourceTypeToResourceModel = {
    term: Term;
    vocabularyList: VocabularyList;
    transcribedAudio: TranscribedAudio;
    book: Book;
    photograph: Photograph;
    spatialFeature: ISpatialFeature;
    bibliographicReference: IBibliographicReference;
    song: Song;
    mediaItem: MediaItem;
};

/**
 * This represents the state of all domain models, excluding their `Connections`
 */
export type InMemorySnapshotOfResources = {
    [K in ResourceType]?: ResourceTypeToResourceModel[K][];
};

export type InMemorySnapshot = {
    resources: InMemorySnapshotOfResources;
    connections: EdgeConnection[];
    tags: Tag[];
    /**
     * We do not intend to leak the abstraction of how the categories are
     * represented in the database here. Defer this to (the db specific) document
     * mapping layer.
     */
    categoryTree: Category[];
    users: CoscradUser[];
    userGroups: CoscradUserGroup[];
    uuids: UuidDocument<AggregateId>[];
};
