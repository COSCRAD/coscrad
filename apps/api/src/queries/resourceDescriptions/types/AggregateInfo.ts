import { AggregateType, IAggregateInfo } from '@coscrad/api-interfaces';

export type AggregateInfo = IAggregateInfo<AggregateType>;
