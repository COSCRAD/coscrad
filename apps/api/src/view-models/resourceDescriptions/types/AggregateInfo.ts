import { AggregateTypeAndDescription } from './AggregateTypeAndDescription';

export type AggregateInfo = AggregateTypeAndDescription & {
    // User-facing
    label: string;
};
