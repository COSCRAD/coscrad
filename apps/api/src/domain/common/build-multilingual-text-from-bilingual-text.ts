import { InternalError } from '../../lib/errors/InternalError';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { buildMultilingualTextWithSingleItem } from './build-multilingual-text-with-single-item';
import {
    MultilingualText,
    MultilingualTextItem,
    MultilingualTextItemRole,
} from './entities/multilingual-text';

export class InvalidBilingualTextException extends InternalError {
    constructor() {
        super(
            `Bilingual text must have non-empty text for either the primary or secondary language`
        );
    }
}

export type TextAndLanguageCode = Pick<MultilingualTextItem, 'text' | 'languageCode'>;

export const buildMultilingualTextFromBilingualText = (
    primary: TextAndLanguageCode,
    secondary: TextAndLanguageCode
) => {
    if (isNullOrUndefined(primary) || isNullOrUndefined(secondary))
        throw new InvalidBilingualTextException();

    if (isNullOrUndefined(primary.text)) {
        if (isNullOrUndefined(secondary.text)) throw new InvalidBilingualTextException();

        return buildMultilingualTextWithSingleItem(secondary.text, secondary.languageCode);
    }

    const multilingualTextForPrimary = buildMultilingualTextWithSingleItem(
        primary.text,
        primary.languageCode
    );

    return isNullOrUndefined(secondary.text)
        ? multilingualTextForPrimary
        : (multilingualTextForPrimary.translate(
              new MultilingualTextItem({
                  ...secondary,
                  role: MultilingualTextItemRole.freeTranslation,
              })
          ) as MultilingualText);
};
