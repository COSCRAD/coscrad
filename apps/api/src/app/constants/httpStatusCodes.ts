import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
