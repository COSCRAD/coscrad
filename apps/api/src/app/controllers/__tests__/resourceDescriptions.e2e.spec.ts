import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { ResourceType } from '../../../domain/types/ResourceType';
import generateDatabaseNameForTestSuite from '../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import { AggregateInfo } from '../../../view-models/resourceDescriptions/types/AggregateInfo';
import httpStatusCodes from '../../constants/httpStatusCodes';
import setUpIntegrationTest from './setUpIntegrationTest';
describe('GET /resources', () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    beforeAll(async () => {
        ({ app } = await setUpIntegrationTest({
            ARANGO_DB_NAME: testDatabaseName,
            GLOBAL_PREFIX: 'testApiPrefix',
        }));
    });

    it('should return a 200', async () => {
        const result = await request(app.getHttpServer()).get('/resources');

        expect(result.status).toBe(httpStatusCodes.ok);
    });

    it('should return one description for each resource type', async () => {
        const result = await request(app.getHttpServer()).get('/resources');

        const body = result.body as AggregateInfo[];

        // TODO [optimization]: avoid loop within loop here
        const isThereAnEntryForEveryResourceType = Object.values(ResourceType).every(
            (resourceType) =>
                body.some(({ type: responseResourceType }) => resourceType === responseResourceType)
        );

        expect(isThereAnEntryForEveryResourceType).toBe(true);
    });

    it('should return the expected result', async () => {
        const result = await request(app.getHttpServer()).get('/resources');

        expect(result.body).toMatchSnapshot();
    });

    afterAll(async () => {
        await app.close();
    });
});
