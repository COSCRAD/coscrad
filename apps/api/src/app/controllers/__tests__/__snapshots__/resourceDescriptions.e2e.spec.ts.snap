// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /resources should return the expected result 1`] = `
Array [
  Object {
    "description": "A term is a word, phrase, or sentence.",
    "label": "Term",
    "link": "/resources/terms",
    "schema": Object {
      "audioURL": Object {
        "coscradDataType": "URL",
        "description": "a web link to a digital audio recording of this term's pronunciation",
        "isArray": false,
        "isOptional": false,
        "label": "audio link",
      },
      "contributor": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "name of the person who contributed this term",
        "isArray": false,
        "isOptional": false,
        "label": "contributor",
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "sourceProject": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the name of the project through which the term was collected",
        "isArray": false,
        "isOptional": true,
        "label": "Source Project",
      },
      "term": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the term in the language",
        "isArray": false,
        "isOptional": true,
        "label": "text (language)",
      },
      "termEnglish": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the text in the colonial language",
        "isArray": false,
        "isOptional": true,
        "label": "text (colonial language)",
      },
    },
    "type": "term",
  },
  Object {
    "description": "A vocabulary list gathers terms with filters that apply within the context of the vocabulary list.",
    "label": "Vocabulary List",
    "link": "/resources/vocabularyLists",
    "schema": Object {
      "entries": Object {
        "complexDataType": "NESTED_TYPE",
        "description": "all terms in this vocabulary list together with their filter properties",
        "isArray": true,
        "isOptional": false,
        "label": "entries",
        "schema": Object {
          "term": Object {
            "complexDataType": "NESTED_TYPE",
            "description": "the term that is included in the vocabulary list via this entry",
            "isArray": false,
            "isOptional": false,
            "label": "term",
            "schema": Object {
              "audioURL": Object {
                "coscradDataType": "URL",
                "description": "a web link to a digital audio recording of this term's pronunciation",
                "isArray": false,
                "isOptional": false,
                "label": "audio link",
              },
              "contributor": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "name of the person who contributed this term",
                "isArray": false,
                "isOptional": false,
                "label": "contributor",
              },
              "id": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "unique identifier",
                "isArray": false,
                "isOptional": false,
                "label": "ID",
              },
              "sourceProject": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the name of the project through which the term was collected",
                "isArray": false,
                "isOptional": true,
                "label": "Source Project",
              },
              "term": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the term in the language",
                "isArray": false,
                "isOptional": true,
                "label": "text (language)",
              },
              "termEnglish": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the text in the colonial language",
                "isArray": false,
                "isOptional": true,
                "label": "text (colonial language)",
              },
            },
          },
        },
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "name": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "name of the vocabulary list in the language",
        "isArray": false,
        "isOptional": true,
        "label": "name (language)",
      },
      "nameEnglish": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "name of the vocabulary list in the colonial language",
        "isArray": false,
        "isOptional": true,
        "label": "name (colonial language)",
      },
    },
    "type": "vocabularyList",
  },
  Object {
    "description": "A transcribed audio item includes a link to an audio recording and the associated transcript.",
    "label": "Transcribed Audio Item",
    "link": "/resources/transcribedAudioItems",
    "schema": Object {
      "audioURL": Object {
        "coscradDataType": "URL",
        "description": "a web link to an accompanying digital audio file",
        "isArray": false,
        "isOptional": false,
        "label": "audio link",
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "lengthMilliseconds": Object {
        "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
        "description": "the length of the accompanying audio file in milliseconds",
        "isArray": false,
        "isOptional": false,
        "label": "length (ms)",
      },
      "mimeType": Object {
        "complexDataType": "ENUM",
        "description": "the media format type",
        "enumLabel": "MIME type",
        "enumName": "MIMEType",
        "isArray": false,
        "isOptional": false,
        "label": "MIME type",
        "labelsAndValues": Array [
          Object {
            "label": "mp3",
            "value": "audio/mpeg",
          },
          Object {
            "label": "mp4",
            "value": "video/mp4",
          },
          Object {
            "label": "png",
            "value": "image/png",
          },
        ],
      },
      "name": Object {
        "coscradDataType": "MULTILINGUAL_TEXT",
        "description": "name of the transcript",
        "isArray": false,
        "isOptional": false,
        "label": "name",
      },
      "text": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "a plain-text representation of the transcript",
        "isArray": false,
        "isOptional": false,
        "label": "plain text",
      },
    },
    "type": "audioItem",
  },
  Object {
    "description": "A book is a digital representation of a text, organized into pages.",
    "label": "Book",
    "link": "/resources/books",
    "schema": Object {
      "author": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the author of the book",
        "isArray": false,
        "isOptional": false,
        "label": "author",
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "publicationDate": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "a free form string that specifies the publication date",
        "isArray": false,
        "isOptional": true,
        "label": "publication date",
      },
      "subtitle": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "subtitle of the book (in whatever language)",
        "isArray": false,
        "isOptional": true,
        "label": "subtitle",
      },
      "title": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "title of the book (in whatever language)",
        "isArray": false,
        "isOptional": false,
        "label": "title",
      },
    },
    "type": "book",
  },
  Object {
    "description": "A photograph is a digital representation of an analog photograph and its metadata.",
    "label": "Photograph",
    "link": "/resources/photographs",
    "schema": Object {
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "imageURL": Object {
        "coscradDataType": "URL",
        "description": "a web link to a digital version of the photograph",
        "isArray": false,
        "isOptional": false,
        "label": "image link",
      },
      "photographer": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the person who took the picture",
        "isArray": false,
        "isOptional": false,
        "label": "photograph",
      },
    },
    "type": "photograph",
  },
  Object {
    "description": "A spatial feature may be a point, line, or polygon on the map.",
    "label": "Spatial Feature",
    "link": "/resources/spatialFeatures",
    "schema": Object {
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
    },
    "type": "spatialFeature",
  },
  Object {
    "description": "A bibliographic reference is a reference to, but not a digital representation of, a research resource.",
    "label": "BibliographicReference",
    "link": "/resources/bibliographicReferences",
    "schema": Object {
      "data": Object {
        "complexDataType": "UNION_TYPE",
        "description": "citation information for this bibliographic reference",
        "discriminantPath": "type",
        "isArray": false,
        "isOptional": false,
        "label": "reference data",
        "schemaDefinitions": Array [
          Object {
            "discriminant": "book",
            "schema": Object {
              "abstract": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "a brief summary of the referenced book",
                "isArray": false,
                "isOptional": true,
                "label": "abstract",
              },
              "creators": Object {
                "complexDataType": "NESTED_TYPE",
                "description": "those responsible for creating this work",
                "isArray": true,
                "isOptional": false,
                "label": "creators",
                "schema": Object {
                  "name": Object {
                    "coscradDataType": "NON_EMPTY_STRING",
                    "description": "full name of work's creator (free-form text)",
                    "isArray": false,
                    "isOptional": false,
                    "label": "name",
                  },
                  "type": Object {
                    "complexDataType": "ENUM",
                    "description": "the person's role in creating the given work",
                    "enumLabel": "Creator Type",
                    "enumName": "BIBLIOGRAPHIC_SUBJECT_CREATOR_TYPE",
                    "isArray": false,
                    "isOptional": false,
                    "label": "creator type",
                    "labelsAndValues": Array [
                      Object {
                        "label": "artist",
                        "value": "artist",
                      },
                      Object {
                        "label": "author",
                        "value": "author",
                      },
                      Object {
                        "label": "director",
                        "value": "director",
                      },
                    ],
                  },
                },
              },
              "isbn": Object {
                "coscradDataType": "ISBN",
                "description": "a standardized globally unique identifier for the referenced book",
                "isArray": false,
                "isOptional": true,
                "label": "ISBN",
              },
              "numberOfPages": Object {
                "coscradDataType": "POSITIVE_INTEGER",
                "description": "the total number of pages in the referenced book",
                "isArray": false,
                "isOptional": true,
                "label": "number of pages",
              },
              "place": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the place where the referenced book was published",
                "isArray": false,
                "isOptional": true,
                "label": "place",
              },
              "publisher": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the publisher of the referenced book",
                "isArray": false,
                "isOptional": true,
                "label": "publisher",
              },
              "title": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the title of the referenced book",
                "isArray": false,
                "isOptional": false,
                "label": "title",
              },
              "url": Object {
                "coscradDataType": "URL",
                "description": "an link to an external digital representation of the book",
                "isArray": false,
                "isOptional": true,
                "label": "external link",
              },
              "year": Object {
                "coscradDataType": "YEAR",
                "description": "A number representing the year of publication",
                "isArray": false,
                "isOptional": true,
                "label": "year",
              },
            },
          },
          Object {
            "discriminant": "courtCase",
            "schema": Object {
              "abstract": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "a brief summary of the court case",
                "isArray": false,
                "isOptional": true,
                "label": "abstract",
              },
              "caseName": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the official name of the court case",
                "isArray": false,
                "isOptional": false,
                "label": "case name",
              },
              "court": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the name of the court in which the case was decided",
                "isArray": false,
                "isOptional": true,
                "label": "court",
              },
              "dateDecided": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the date on which the court case was decided",
                "isArray": false,
                "isOptional": true,
                "label": "date decided",
              },
              "pages": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "summary information about the pages relevant to this court case",
                "isArray": false,
                "isOptional": true,
                "label": "pages",
              },
              "url": Object {
                "coscradDataType": "URL",
                "description": "external link to digital materials regarding this court case",
                "isArray": false,
                "isOptional": true,
                "label": "external link",
              },
            },
          },
          Object {
            "discriminant": "journalArticle",
            "schema": Object {
              "abstract": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "a brief summary of the journal article",
                "isArray": false,
                "isOptional": true,
                "label": "abstract",
              },
              "creators": Object {
                "complexDataType": "NESTED_TYPE",
                "description": "the authors of the referenced journal article",
                "isArray": true,
                "isOptional": false,
                "label": "creators",
                "schema": Object {
                  "name": Object {
                    "coscradDataType": "NON_EMPTY_STRING",
                    "description": "full name of work's creator (free-form text)",
                    "isArray": false,
                    "isOptional": false,
                    "label": "name",
                  },
                  "type": Object {
                    "complexDataType": "ENUM",
                    "description": "the person's role in creating the given work",
                    "enumLabel": "Creator Type",
                    "enumName": "BIBLIOGRAPHIC_SUBJECT_CREATOR_TYPE",
                    "isArray": false,
                    "isOptional": false,
                    "label": "creator type",
                    "labelsAndValues": Array [
                      Object {
                        "label": "artist",
                        "value": "artist",
                      },
                      Object {
                        "label": "author",
                        "value": "author",
                      },
                      Object {
                        "label": "director",
                        "value": "director",
                      },
                    ],
                  },
                },
              },
              "doi": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "an alternative unique identifier in the form of a permalink",
                "isArray": false,
                "isOptional": true,
                "label": "DOI",
              },
              "issn": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "a standardized unique identifier for the journal article",
                "isArray": false,
                "isOptional": true,
                "label": "ISSN",
              },
              "issueDate": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "unstructured text representation of the date the journal article was issued",
                "isArray": false,
                "isOptional": false,
                "label": "issue date",
              },
              "publicationTitle": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the name of the journal in which the article was published",
                "isArray": false,
                "isOptional": true,
                "label": "publication title",
              },
              "title": Object {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "the title of the referenced journal article",
                "isArray": false,
                "isOptional": false,
                "label": "title",
              },
              "url": Object {
                "coscradDataType": "URL",
                "description": "an external link to a digital representation of the referenced journal article",
                "isArray": false,
                "isOptional": true,
                "label": "external link",
              },
            },
          },
        ],
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
    },
    "type": "bibliographicReference",
  },
  Object {
    "description": "A song includes a link to an audio recording along with metadata and lyrics (when available).",
    "label": "Song",
    "link": "/resources/songs",
    "schema": Object {
      "audioURL": Object {
        "coscradDataType": "URL",
        "description": "a web link to the digital audio file for playback",
        "isArray": false,
        "isOptional": false,
        "label": "audio link",
      },
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "lengthMilliseconds": Object {
        "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
        "description": "length of the audio file in milliseconds",
        "isArray": false,
        "isOptional": false,
        "label": "length (ms)",
      },
      "lyrics": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the lyrics of the song",
        "isArray": false,
        "isOptional": true,
        "label": "lyrics",
      },
      "startMilliseconds": Object {
        "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
        "description": "the starting timestamp for the audio file",
        "isArray": false,
        "isOptional": false,
        "label": "start (ms)",
      },
      "title": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the title of the song in the language",
        "isArray": false,
        "isOptional": true,
        "label": "title",
      },
      "titleEnglish": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "the title of the song in the colonial language",
        "isArray": false,
        "isOptional": true,
        "label": "title (colonial language)",
      },
    },
    "type": "song",
  },
  Object {
    "description": "A media item is a digital representation of an audio or video recording.",
    "label": "Media Item",
    "link": "/resources/mediaItems",
    "schema": Object {
      "id": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "unique identifier",
        "isArray": false,
        "isOptional": false,
        "label": "ID",
      },
      "lengthMilliseconds": Object {
        "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
        "description": "length of the media item in milliseconds",
        "isArray": false,
        "isOptional": false,
        "label": "length (ms)",
      },
      "mimeType": Object {
        "complexDataType": "ENUM",
        "description": "technical specification of the format of the media item",
        "enumLabel": "MIME type",
        "enumName": "MIMEType",
        "isArray": false,
        "isOptional": false,
        "label": "MIME type",
        "labelsAndValues": Array [
          Object {
            "label": "mp3",
            "value": "audio/mpeg",
          },
          Object {
            "label": "mp4",
            "value": "video/mp4",
          },
          Object {
            "label": "png",
            "value": "image/png",
          },
        ],
      },
      "title": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "title of the media item in the language",
        "isArray": false,
        "isOptional": true,
        "label": "title (language)",
      },
      "titleEnglish": Object {
        "coscradDataType": "NON_EMPTY_STRING",
        "description": "title of the media item in the colonial language",
        "isArray": false,
        "isOptional": true,
        "label": "title (colonial language)",
      },
      "url": Object {
        "coscradDataType": "URL",
        "description": "a web link to the corresponding media file",
        "isArray": false,
        "isOptional": false,
        "label": "url",
      },
    },
    "type": "mediaItem",
  },
]
`;
